### PR TITLE
(PUP-10391) Quiet Ruby 2.7 URI.escape warning

### PR DESCRIPTION
**Pull request by melissa**

### CODE REVIEW

File: lib/puppet/util.rb [hunk 1]
```diff
@@ -479,6 +487,11 @@ def uri_encode(path, opts = { :allow_fragment => false })
   end
   module_function :uri_encode
 
+  def uri_unescape(path)
+    URI.unescape(path)
```
Comment (joshcooper) [inline lib/puppet/util.rb:491]: Does the `URI.unescape` need the `$VERBOSE` treatment too? Or maybe ruby only deprecated `escape`?

Comment (melissa) [reply lib/puppet/util.rb:493]: Yes, yes it does.
```
irb(main):004:0> require 'URI'
=> true
irb(main):005:0> URI.escape('path to')
(irb):5: warning: URI.escape is obsolete
=> "path%20to"
irb(main):007:0> URI.unescape('path%20to')
(irb):7: warning: URI.unescape is obsolete
=> "path to"
```

Comment (melissa) [reply lib/puppet/util.rb:493]: Updated

**Comment by melissa** [comment]
@joshcooper there's also the use of `URI.decode`, which throws an error. Should I wrap that as well?
```
irb(main):004:0> URI.decode('aspth')
(irb):4: warning: URI.unescape is obsolete
```
```
lib/puppet/forge.rb:    URI.decode(uri.tr('+', ' '))
```
Apparently `URI.decode` calls `URI.unescape`.

**Comment by joshcooper** [comment]
> Should I wrap that as well?

Yeah, ruby defines `decode` as an alias of `unescape` or the other way around, so 👍 to wrapping both.

Also one nit, if the URI.unescape or URI.escape method raise, then we never restore the old $VERBOSE level, so best to wrap with begin/rescue:

```ruby
$VERBOSE = false
begin
  URI.unescape(..)
ensure
  $VERBOSE = orig_verbose unless Puppet::Util::Platform.jruby?
end
```

And if you do that, then you can do `return URI.unescape(..)` from within the begin/rescue and its return value will be returned from the `P::U.uri_encode` method. Same for uri_decode.

**Comment by melissa** [comment]
Updated!

**Comment by joshcooper** [comment]
Looks good, just had one nit, I'm still getting warnings due to the spec files using URI directly like:

```
$ bx rspec spec/unit/type/file/source_spec.rb
...
/Users/josh/work/puppet/spec/unit/type/file/source_spec.rb:25: warning: URI.unescape is obsolete
```

It'd be handy to have the specs call `Puppet::Util.uri_encode` to reduce the noise

**Comment by melissa** [comment]
Whoops... sorry about that. Updated!

**Comment by melissa** [comment]
```
Failures:
  1) puppet agent multiple agents running exits if maxwaitforlock is exceeded
     Failure/Error:
       expect {
         agent.command_line.args << '--test'
         agent.run
       }.to exit_with(1).and output(/Exiting now because the maxwaitforlock timeout has been exceeded./).to_stdout
       expected block to output /Exiting now because the maxwaitforlock timeout has been exceeded./ to stdout, but output "\e[0;32mInfo: Another puppet instance is already running; --waitforlock flag used, waiting for runni... 1 seconds.\e[0m\n\e[0;32mInfo: Retrieving pluginfacts\e[0m\n\e[0;32mInfo: Retrieving plugin\e[0m\n"
     # ./spec/integration/application/agent_spec.rb:386:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:180:in `block (2 levels) in <top (required)>'
     # util/rspec_runner:44:in `run'
     # util/rspec_runner:59:in `<main>'
Finished in 1 minute 45.14 seconds (files took 6.27 seconds to load)
Finished in 14 minutes 35 seconds
26048 examples, 1 failure, 83 pending
```
Looks like a race condition? But It keeps popping up, so I'm not sure how race-y this failure is

**Comment by melissa** [comment]
Oh, cool, tests are passing now

**Comment by joshcooper** [comment]
I ran the tests with `bx rspec -fd spec` on ruby 2.7 and didn't see any URI deprecation warnings.

**Comment by anatol** [comment]
Would it be better to migrate this `escape` code to a long-term solution instead?

With ruby 3.0 released several months ago this `escape` warning became error as there is no `URI.escape()` function anymore.

**Pull request merged**